### PR TITLE
Plan quality metrics

### PR DESCRIPTION
**Pull request by alvalentini**

### CODE REVIEW

File: unified_planning/io/python_writer.py [hunk 1]
```diff
@@ -47,7 +47,7 @@ def walk_forall(self, expression, args):
     def walk_variable_exp(self, expression, args):
         assert len(args) == 0
         v = expression.variable()
-        return f'up.model.Variable("{v.name()}", {_print_python_type(v.type())})'
+        return f'emgr.VariableExp(up.model.Variable("{v.name()}", {_print_python_type(v.type())}))'
```
Comment (mikand) [inline unified_planning/io/python_writer.py:50]: Was this a bug of the python printer? How is it not caught by a test?

Comment (mikand) [inline unified_planning/io/python_writer.py:50]: Was this a bug of the python printer? How is it not caught by a test?

File: unified_planning/io/python_writer.py [hunk 2]
```diff
@@ -73,17 +73,17 @@ def walk_fluent_exp(self, expression, args):
         fluent = expression.fluent()
         if args:
             return f'fluent_{fluent.name()}({", ".join(args)})'
-        return f'fluent_{fluent.name()}'
+        return f'emgr.FluentExp(fluent_{fluent.name()})'
```
Comment (mikand) [inline unified_planning/io/python_writer.py:76]: Same as above

Comment (mikand) [inline unified_planning/io/python_writer.py:76]: Same as above

File: unified_planning/model/metrics.py [hunk 3]
```diff
@@ -14,15 +14,28 @@
...
 import unified_planning as up
+from typing import Dict
 
 
 class PlanQualityMetric():
     '''This is the base class of any metric for plan quality'''
     pass
 
 class MinimizeActionCosts(PlanQualityMetric):
+    def __init__(self, costs: Dict['up.model.Action', 'up.model.FNode']):
```
Comment (mikand) [inline unified_planning/model/metrics.py:25]: Shall we add a getter method like `get_action_cost(a:Action) -> FNode`?

Comment (mikand) [inline unified_planning/model/metrics.py:25]: Shall we add a getter method like `get_action_cost(a:Action) -> FNode`?

**Review by mikand — CHANGES_REQUESTED** [review]
@alvalentini Excellent work. I left some comments for consideration. 
@Framba-Luca have a look at the comments on the python printer to see if we need more tests

File: unified_planning/io/pddl_reader.py [hunk 4]
```diff
@@ -523,24 +523,36 @@ def parse_problem(self, domain_filename: str,
...
-                        problem.add_quality_metric(up.model.metrics.MinimizeExpressionOnFinalState(metric_exp))
-                    elif optimization == 'maximize':
-                        problem.add_quality_metric(up.model.metrics.MaximizeExpressionOnFinalState(metric_exp))
+                    metric_exp = self._parse_exp(problem, None, types_map, {}, metric)
+                    if has_actions_cost and optimization == 'minimize' and metric_exp == self._totalcost:
+                        costs = {}
+                        problem._fluents.remove(self._totalcost.fluent())
+                        problem._initial_value.pop(self._totalcost)
+                        use_plan_length = all(False for _ in problem.durative_actions())
+                        for a in problem.instantaneous_actions():
```
[the remark this answers is not in the document]
Comment (alvalentini) [reply unified_planning/io/pddl_reader.py:535]: We do not support costs for durative actions. If durative actions have effects on a `total-cost` fluent, we use the `MinimizeExpressionOnFinalState` metric

File: unified_planning/io/pddl_reader.py [hunk 5]
```diff
@@ -523,24 +523,36 @@ def parse_problem(self, domain_filename: str,
...
+                            cost = None
+                            for e in a.effects():
+                                if e.fluent() == self._totalcost:
+                                    cost = e
+                                    break
```
[the remark this answers is not in the document]
Comment (alvalentini) [reply unified_planning/io/pddl_reader.py:540]: We set the cost only for the actions that have an effect `(increase total-cost <constant>)`.

File: unified_planning/model/metrics.py (hunk 3 above)
Comment (alvalentini) [reply unified_planning/model/metrics.py:25]: Yes, we can add it.

File: unified_planning/grpc/unified_planning.proto [hunk 6]
```diff
@@ -289,6 +285,40 @@ message Assignment {
     Expression value = 2;
 }
 
+
+message MetricBody {
```
[the remark this answers is not in the document]
Comment (alvalentini) [reply unified_planning/grpc/unified_planning.proto:289]: I think both possibilities are good

File: unified_planning/io/python_writer.py (hunk 1 above)
Comment (alvalentini) [reply unified_planning/io/python_writer.py:50]: Yes, it was a bug. 
@Framba-Luca can you answer the question `How is it not caught by a test?`

File: unified_planning/io/python_writer.py (hunk 2 above)
Comment (alvalentini) [reply unified_planning/io/python_writer.py:76]: @Framba-Luca

File: unified_planning/io/pddl_reader.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (alvalentini) [reply unified_planning/io/pddl_reader.py:535]: I mean that we do not support the creation of the `MinimizeActionCosts` metric through the pddl reader when the `total-cost` fluent is used in durative actions

File: unified_planning/grpc/unified_planning.proto (hunk 6 above)
[the remark this answers is not in the document]
Comment (alvalentini) [reply unified_planning/grpc/unified_planning.proto:289]: I prefer the current message structure, where the `MetricBody` is an optional field of the `Metric` message.
But also your suggestion works fine, let me know if you want that I change the proto file according with your suggestion or if we keep it.

File: unified_planning/io/python_writer.py (hunk 1 above)
Comment (mikand) [reply unified_planning/io/python_writer.py:50]: @alvalentini @Framba-Luca  Not sure to follow, is/was this a bug or not? If it was then we need a test that is failing with the previous version and is fixed now, if it was not a bug, that we need justification for this change.

Comment (mikand) [reply unified_planning/io/python_writer.py:50]: Same also for the other changes below

**Review by mikand — APPROVED** [review]
LGTM except for the point on the changes in the python printer. Either add the missing no-regression test or comment on why the changes are needed.

After that, you can go ahead and merge.

**Pull request merged**